src/server/serve.rs

```diff
@@ -53,25 +58,49 @@ pub enum PathType {
     SymlinkFile,
 }
 
+#[derive(Clone)]
+struct MakeService {
+    args: Arc<Args>,
+    gitignore: Arc<Gitignore>,
+}
+
+impl<T> Service<T> for MakeService {
+    type Response = InnerService;
+    type Error = hyper::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, _: T) -> Self::Future {
+        let args = self.args.clone();
+        let gitignore = self.gitignore.clone();
+        let fut = async move { Ok(InnerService { args, gitignore }) };
+        Box::pin(fut)
+    }
+}
+
+impl MakeService {
+    pub fn new(args: Args) -> Self {
+        let args = Arc::new(args);
+        let gitignore = Arc::new(Gitignore::new(args.path.join(".gitignore")).0);
+        Self { args, gitignore }
+    }
+}
+
 /// Run the server.
 pub async fn serve(args: Args) -> BoxResult<()> {
     let address = args.address()?;
     let path_prefix = args.path_prefix.clone().unwrap_or_default();
-    let inner = Arc::new(InnerService::new(args));
-    let make_svc = make_service_fn(move |_| {
-        let inner = inner.clone();
-        async {
-            Ok::<_, Infallible>(service_fn(move |req| {
-                let inner = inner.clone();
-                inner.call(req)
-            }))
-        }
-    });
 
-    let server = hyper::Server::try_bind(&address)?.serve(make_svc);
+    let server = hyper::Server::bind(&address).serve(MakeService::new(args));
     let address = server.local_addr();
-    eprintln!("Files served on http://{}{}", address, path_prefix);
-    server.await?;
+    eprintln!("Files served on http://{address}{path_prefix}");
+
+    if let Err(e) = server.await {
+        eprintln!("Server error: {e:?}");
+    };
```

Comment:
Something forces you to match `Err(…)` here? I remember top level `fn main()` already handles this case.

Reply:
reverted on 481b1fb